file_store.go

```diff
@@ -0,0 +1,208 @@
+package raft
+
+import (
+	"encoding/gob"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+// FileStore implements the LogStore and StableStore interface.
+type FileStore struct {
+	l         sync.RWMutex
+	lowIndex  uint64
+	highIndex uint64
+	dir       string
+}
+
+type uint64Slice []uint64
+
+func (p uint64Slice) Len() int           { return len(p) }
+func (p uint64Slice) Less(i, j int) bool { return p[i] < p[j] }
+func (p uint64Slice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+func NewFileStore(dir string) (*FileStore, error) {
+	if err := os.MkdirAll(filepath.Join(dir, "logs"), 0700); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(filepath.Join(dir, "stable"), 0700); err != nil {
+		return nil, err
+	}
+
+	return &FileStore{
+		dir: dir,
+	}, nil
+}
+
+// GetAll returns all indexes that are currently present in the log store. This
+// is NOT part of the LogStore interface — we use it when snapshotting.
+func (s *FileStore) GetAll() ([]uint64, error) {
```

Comment:
True, this is equal to just ranging over `FirstIndex()` to `LastIndex()`. Only realized that afterwards :).